qa/lock/databricks/databricks_backend.go

```diff
@@ -0,0 +1,176 @@
+package databricks
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"time"
+
+	"github.com/databricks/databricks-sdk-go"
+	"github.com/databricks/databricks-sdk-go/apierr"
+	"github.com/databricks/databricks-sdk-go/qa/lock/core"
+	"github.com/databricks/databricks-sdk-go/service/workspace"
+)
+
+// Backend stores locks in the Databricks workspace. All locks are stored
+// in a fixed directory, and each lock is stored in a file named after the lock ID.
+// The workspace client is shared by all locks.
+type Backend struct {
+	lockClient *databricks.WorkspaceClient
+	lockDir    string
+	lockName   string
+}
+
+func getEnvOrFail(key string) string {
+	val := os.Getenv(key)
+	if val == "" {
+		panic(fmt.Sprintf("%s must be set", key))
+	}
+	return val
+}
+
+// Configure the backend by setting the workspace client and lock file.
+func (d *Backend) PrepareBackend(ctx context.Context, lockId string) error {
+	configJson := getEnvOrFail("LOCK_CLIENT_CONFIG")
+	lockDirectory := getEnvOrFail("LOCK_DIRECTORY")
+
+	c := &databricks.Config{}
+	err := json.Unmarshal([]byte(configJson), c)
+	if err != nil {
+		panic(fmt.Errorf("failed to unmarshal config: %w", err))
+	}
+
+	w, err := databricks.NewWorkspaceClient(c)
+	if err != nil {
+		panic(fmt.Errorf("failed to create workspace client: %w", err))
+	}
+
+	d.lockClient = w
+	d.lockDir = lockDirectory
+	d.lockName = lockId
+	return nil
+}
+
+func (d *Backend) getLockPath() string {
+	return fmt.Sprintf("%s/%s.lock", d.lockDir, d.lockName)
+}
+
+func (d *Backend) getCurrentLockState(ctx context.Context) (*core.LockState, error) {
+	resp, err := d.lockClient.Workspace.Download(ctx, d.getLockPath())
+	if errors.Is(err, apierr.ErrNotFound) {
+		return nil, nil
+	} else if err != nil {
+		return nil, fmt.Errorf("failed to read lock %s state: %w", d.lockName, err)
+	}
+	defer resp.Close()
+	bs, err := io.ReadAll(resp)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read lock %s state: %w", d.lockName, err)
+	}
+
+	bs = bytes.TrimPrefix(bs, []byte("# Databricks notebook source\n"))
+
+	var state core.LockState
+	err = json.Unmarshal(bs, &state)
+	if err != nil {
+		return nil, fmt.Errorf("failed to unmarshal lock %s state: %w", d.lockName, err)
+	}
+
+	return &state, nil
+}
+
+func (d *Backend) uploadLockState(ctx context.Context, contents *core.LockState, overwrite bool) error {
+	buf, err := json.Marshal(contents)
+	if err != nil {
+		return fmt.Errorf("failed to marshal contents: %w", err)
+	}
+
+	opts := []workspace.UploadOption{
+		workspace.UploadLanguage(workspace.LanguagePython),
```

Comment:
Can we omit the language setting here?

Reply:
Nope, otherwise the API thinks this is a ZIP file (probably DBC archive).